File: tianshu/core/rag/store.py
from __future__ import annotations

import json
import re
import threading
import time
import uuid
from dataclasses import dataclass, field

from tianshu.config import WORKSPACE_DIR

RAG_DIR = WORKSPACE_DIR / ".ts-rag"
INDEX_FILE = RAG_DIR / "index.json"
CHUNK_SIZE = 500
CHUNK_OVERLAP = 60
_lock = threading.Lock()


@dataclass
class Chunk:
    text: str
    doc_id: str
    version: int
    seq: int


@dataclass
class DocumentVersion:
    version: int
    ts: str
    source: str
    chunks: list[str] = field(default_factory=list)
    superseded: bool = False


@dataclass
class Document:
    id: str
    title: str
    versions: list[DocumentVersion] = field(default_factory=list)

    @property
    def latest(self) -> DocumentVersion | None:
        return self.versions[-1] if self.versions else None


def _ensure() -> None:
    RAG_DIR.mkdir(parents=True, exist_ok=True)
    if not INDEX_FILE.exists():
        INDEX_FILE.write_text("{}", encoding="utf-8")


def _load() -> dict:
    _ensure()
    try:
        return json.loads(INDEX_FILE.read_text(encoding="utf-8"))
    except (ValueError, OSError):
        return {}


def _save(data: dict) -> None:
    _ensure()
    INDEX_FILE.write_text(json.dumps(data, ensure_ascii=False, indent=1), encoding="utf-8")


def _split_chunks(text: str) -> list[str]:
    text = re.sub(r"\n{3,}", "\n\n", text).strip()
    if not text:
        return []
    paragraphs = re.split(r"\n\s*\n", text)
    chunks: list[str] = []
    buf = ""
    for para in paragraphs:
        if len(buf) + len(para) + 2 <= CHUNK_SIZE:
            buf = f"{buf}\n\n{para}".strip() if buf else para
            continue
        if buf:
            chunks.append(buf)
        if len(para) > CHUNK_SIZE:
            words = re.findall(r"\S+", para)
            cur = ""
            for w in words:
                if len(cur) + len(w) + 1 > CHUNK_SIZE:
                    chunks.append(cur)
                    tail = " ".join(cur.split()[-CHUNK_OVERLAP:]) if CHUNK_OVERLAP else ""
                    cur = f"{tail} {w}".strip()
                else:
                    cur = f"{cur} {w}".strip()
            if cur:
                chunks.append(cur)
        else:
            buf = para
    if buf:
        chunks.append(buf)
    return chunks


def ingest_text(text: str, source: str, doc_id: str = "", title: str = "") -> str:
    with _lock:
        data = _load()
        if not doc_id:
            doc_id = uuid.uuid4().hex[:12]
        doc = data.setdefault("docs", {}).setdefault(doc_id, {"id": doc_id, "title": title or source, "versions": []})
        if title and not doc.get("title"):
            doc["title"] = title
        for v in doc["versions"]:
            v["superseded"] = True
        version = len(doc["versions"]) + 1
        entry: dict = {
            "version": version,
            "ts": time.strftime("%Y-%m-%d %H:%M:%S"),
            "source": source,
            "chunks": _split_chunks(text),
            "superseded": False,
        }
        doc["versions"].append(entry)
        _save(data)
    return f"{doc_id} v{version}"



def _latest_chunks(doc: dict) -> list[Chunk]:
    out: list[Chunk] = []
    for v in doc.get("versions", []):
        if v.get("superseded"):
            continue
        for i, text in enumerate(v.get("chunks", [])):
            out.append(Chunk(text=text, doc_id=doc["id"], version=v.get("version", 1), seq=i))
    return out


def all_chunks() -> list[Chunk]:
    data = _load()
    out: list[Chunk] = []
    for doc in data.get("docs", {}).values():
        out.extend(_latest_chunks(doc))
    return out


def get_doc(doc_id: str) -> Document | None:
    data = _load()
    raw = data.get("docs", {}).get(doc_id)
    if not raw:
        return None
    versions = [
        DocumentVersion(
            version=v.get("version", 1),
            ts=v.get("ts", ""),
            source=v.get("source", ""),
            chunks=v.get("chunks", []),
            superseded=bool(v.get("superseded")),
        )
        for v in raw.get("versions", [])
    ]
    return Document(id=doc_id, title=raw.get("title", doc_id), versions=versions)


def list_docs() -> str:
    data = _load()
    rows = []
    for doc in data.get("docs", {}).values():
        latest = max((v.get("version", 1) for v in doc.get("versions", [])), default=0)
        rows.append(f"{doc['id']}  {doc.get('title','')}  (v{latest})")
    return "\n".join(rows) if rows else "(文档库为空)"

File: tiankui-ts/src/core/rag/store.ts
import * as fs from "node:fs";
import * as path from "node:path";
import { randomUUID } from "node:crypto";
import { WORKSPACE_DIR } from "../../config.js";

export const RAG_DIR = process.env.TIANKUI_RAG_DIR
  ? path.resolve(process.env.TIANKUI_RAG_DIR)
  : path.join(WORKSPACE_DIR, ".ts-rag");
export const INDEX_FILE = path.join(RAG_DIR, "index.json");
export const CHUNK_SIZE = 500;
export const CHUNK_OVERLAP = 60;

export interface Chunk {
  text: string;
  doc_id: string;
  version: number;
  seq: number;
}

export interface DocVersion {
  version: number;
  ts: string;
  source: string;
  chunks: string[];
  superseded: boolean;
}

export interface Doc {
  id: string;
  title: string;
  versions: DocVersion[];
}

function ensure(): void {
  fs.mkdirSync(RAG_DIR, { recursive: true });
  if (!fs.existsSync(INDEX_FILE)) {
    fs.writeFileSync(INDEX_FILE, "{}", "utf-8");
  }
}

function load(): Record<string, unknown> {
  ensure();
  try {
    return JSON.parse(fs.readFileSync(INDEX_FILE, "utf-8"));
  } catch {
    return {};
  }
}

function save(data: Record<string, unknown>): void {
  ensure();
  fs.writeFileSync(INDEX_FILE, JSON.stringify(data, null, 1), "utf-8");
}

export function splitChunks(text: string): string[] {
  text = text.replace(/\n{3,}/g, "\n\n").trim();
  if (!text) return [];
  const paragraphs = text.split(/\n\s*\n/);
  const chunks: string[] = [];
  let buf = "";
  for (const para of paragraphs) {
    if (buf.length + para.length + 2 <= CHUNK_SIZE) {
      buf = buf ? `${buf}\n\n${para}`.trim() : para;
      continue;
    }
    if (buf) chunks.push(buf);
    if (para.length > CHUNK_SIZE) {
      const words = para.match(/\S+/g) ?? [];
      let cur = "";
      for (const w of words) {
        if (cur.length + w.length + 1 > CHUNK_SIZE) {
          chunks.push(cur);
          const tail = CHUNK_OVERLAP ? cur.split(" ").slice(-CHUNK_OVERLAP).join(" ") : "";
          cur = `${tail} ${w}`.trim();
        } else {
          cur = `${cur} ${w}`.trim();
        }
      }
      if (cur) chunks.push(cur);
    } else {
      buf = para;
    }
  }
  if (buf) chunks.push(buf);
  return chunks;
}

export function ingestText(text: string, source: string, docId = "", title = ""): string {
  const data = load();
  let docs = data["docs"] as Record<string, unknown>;
  if (!docs) {
    docs = {};
    data["docs"] = docs;
  }
  if (!docId) docId = randomUUID().replace(/-/g, "").slice(0, 12);
  let doc = docs[docId] as Record<string, unknown>;
  if (!doc) {
    doc = { id: docId, title: title || source, versions: [] };
    docs[docId] = doc;
  }
  if (title && !doc["title"]) doc["title"] = title;
  const versions = (doc["versions"] as Record<string, unknown>[]) ?? [];
  for (const v of versions) v["superseded"] = true;
  const version = versions.length + 1;
  const ts = new Date().toISOString().slice(0, 19).replace("T", " ");
  versions.push({
    version,
    ts,
    source,
    chunks: splitChunks(text),
    superseded: false,
  });
  doc["versions"] = versions;
  save(data);
  return `${docId} v${version}`;
}

function latestChunks(doc: Record<string, unknown>): Chunk[] {
  const out: Chunk[] = [];
  for (const v of (doc["versions"] as Record<string, unknown>[]) ?? []) {
    if (v["superseded"]) continue;
    const chunks = (v["chunks"] as string[]) ?? [];
    chunks.forEach((text, i) => {
      out.push({ text, doc_id: String(doc["id"]), version: Number(v["version"] ?? 1), seq: i });
    });
  }
  return out;
}

export function allChunks(): Chunk[] {
  const data = load();
  const out: Chunk[] = [];
  for (const doc of Object.values((data["docs"] as Record<string, unknown>) ?? {})) {
    out.push(...latestChunks(doc as Record<string, unknown>));
  }
  return out;
}

export function getDoc(docId: string): Doc | null {
  const data = load();
  const docs = (data["docs"] as Record<string, unknown>) ?? {};
  const raw = docs[docId] as Record<string, unknown> | undefined;
  if (!raw) return null;
  const versions: DocVersion[] = ((raw["versions"] as Record<string, unknown>[]) ?? []).map((v) => ({
    version: Number(v["version"] ?? 1),
    ts: String(v["ts"] ?? ""),
    source: String(v["source"] ?? ""),
    chunks: (v["chunks"] as string[]) ?? [],
    superseded: Boolean(v["superseded"]),
  }));
  return { id: docId, title: String(raw["title"] ?? docId), versions };
}

export function listDocs(): string {
  const data = load();
  const docs = (data["docs"] as Record<string, unknown>) ?? {};
  const rows = Object.values(docs).map((doc) => {
    const d = doc as Record<string, unknown>;
    const versions = (d["versions"] as Record<string, unknown>[]) ?? [];
    const latest = versions.reduce((m, v) => Math.max(m, Number(v["version"] ?? 1)), 0);
    return `${d["id"]}  ${d["title"] ?? ""}  (v${latest})`;
  });
  return rows.length ? rows.join("\n") : "(文档库为空)";
}
